### PR TITLE
Convert parsed table caption to HTML to keep formatting

### LINKED ISSUE #2573: Markdown styling in table captions
### Bug description

Quarto does not seem to allow any styling of the text in table captions like _italics_ etc.
This problem does not occur with figure captions.

### Screenshot
image

### .qmd file (can't attach):
````
---
title: "No markdown styling in table caption"
format:
  html:
    theme: default
    keep-md: true
---
```{r}
# captions for figure and table with md styling
caption_table <- "Can't use ^superscript^ or _italics_ in table caption"
caption_figure <- "Can use ^superscript^ or _italics_ in figure caption"
```


```{r}
#| tbl-cap:  !expr caption_table
# caption is set in qmd with: #| tbl-cap: !expr caption_table
library(kableExtra)
column_spec(kable(head(cars)),column = 1:2, width = "15em")
```


```{r}
#| fig-cap: !expr caption_figure
# caption is set in qmd with: #| tbl-cap: !expr caption_figure
plot(cars)
```

```{r}
sessionInfo()
```


````

Table captions are put inside some sort of div, which ignores markdown synthax (or latex syntax when rendering to pdf), for this reason I included the intermediate .md file.

###  .md file excerpt (also see attachment):
````
::: {.cell tbl-cap='Can\'t use ^superscript^ or _italics_ in table caption'}

```{.r .cell-code}
# caption is set in qmd with: #| tbl-cap: !expr caption_table
library(kableExtra)
column_spec(kable(cars),column = 1:2, width = "15em")
```

::: {.cell-output-display}
....
````
[reprex_table_formatting.md](https://github.com/quarto-dev/quarto-cli/files/9632422/reprex_table_formatting.md)

### Info
Rstudio IDE: 2021.09.0 Build 351

Session info:
```
R version 4.1.3 (2022-03-10)
Platform: x86_64-re
dhat-linux-gnu (64-bit)
Running under: Red Hat Enterprise Linux 8.5 (Ootpa)

Matrix products: default
BLAS/LAPACK: /usr/lib64/libopenblas-r0.3.15.so

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
 [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
 [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
 [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
 [9] LC_ADDRESS=C               LC_TELEPHONE=C            
[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] kableExtra_1.3.4

loaded via a namespace (and not attached):
 [1] rstudioapi_0.13   knitr_1.39        xml2_1.3.3        magrittr_2.0.3   
 [5] rvest_1.0.2       munsell_0.5.0     viridisLite_0.4.0 colorspace_2.0-3 
 [9] R6_2.5.1          rlang_1.0.5       fastmap_1.1.0     highr_0.9        
[13] stringr_1.4.0     httr_1.4.3        tools_4.1.3       webshot_0.5.3    
[17] xfun_0.31         cli_3.3.0         systemfonts_1.0.4 htmltools_0.5.2  
[21] yaml_2.3.5        digest_0.6.29     lifecycle_1.0.1   htmlwidgets_1.5.4
[25] glue_1.6.2        evaluate_0.15     rmarkdown_2.14    stringi_1.7.6    
[29] compiler_4.1.3    scales_1.2.0      jsonlite_1.8.0    svglite_2.1.0    ****
```

### Checklist

- [X] Please include a minimal, fully reproducible example in a single .qmd file? Please provide the whole file rather than the snippet you believe is causing the issue.
- [X] Please [format your issue](https://quarto.org/bug-reports.html#formatting-make-githubs-markdown-work-for-us) so it is easier for us to read the bug report.
- [X] Please document the RStudio IDE version you're running (if applicable), by providing the value displayed in the "About RStudio" main menu dialog?
- [X] Please document the operating system you're running. If on Linux, please provide the specific distribution.

### PR DESCRIPTION
**Pull request by cderv**
This fixes #2573 

I am using `stringEscape` here because this is our helper to do this string conversion, but it is poorly name now. 
It was introduced initially to get correct escaping in LaTeX, but seems to be used more broadly now. 

By doing this change, I correctly get this in the HTML
````
<caption><p>Can’t use <sup>superscript</sup> or <em>italics</em> in table
caption</p></caption>
````

which looks like this 
image

It seems ok to do that here, we do such handling in other place like for LaTeX tables, or callouts.

### CODE REVIEW

**Comment by cscheid** [comment]
Should we be checking if the output format is PDF, HTML, Docx, etc and using different code for each?

**Comment by cderv** [comment]
I am inside raw HTML table here so I assumed we are in HTML output
https://github.com/quarto-dev/quarto-cli/pull/2583/files#diff-8a36d99e59ea99952e696c61ae5f36a045c0a8538270020bf1cc315d5e41f022L158

I add missed the other HTML table so I added that. 

We do the same for LaTeX table in `applyLatexTableCaption()`

But we could also use `FORMAT` pandoc variable maybe here, to target the correct format 🤔 

For DOCX, I did not check what we do.

**Comment by cderv** [comment]
> For DOCX, I did not check what we do.

I don't think we handle specifically table in docx written using ```` ```{=openxml} ```` syntax. 

It currently works with markdown table, and html table also with this PR.

**Comment by patrickvdb** [comment]
What would be the reason to use the `FORMAT` pandoc variable over applying the the same solution for `applyLatexTableCaption()`?

Here a diff in a local commit that works for me at the moment.

````
````

**Comment by patrickvdb** [comment]
What would be the advantage of using `FORMAT` over applying the same solution in `applyLatexTableCaption()`? The following change works for my current PDF reports.

````
diff --git a/src/resources/filters/quarto-pre/table-captions.lua b/src/resources/filters/quarto-pre/table-captions.lua
index 6ae00a1df..0691007ad 100644
--- a/src/resources/filters/quarto-pre/table-captions.lua
+++ b/src/resources/filters/quarto-pre/table-captions.lua
@@ -209,11 +209,8 @@ function applyLatexTableCaption(latex, tblCaption, tblLabel, tablePattern)
   end
   -- apply table caption and label
   local beginCaption, captionText, endCaption = latex:match(latexCaptionPattern)
-  if #tblCaption > 0 then
-    captionText = pandoc.utils.stringify(tblCaption)
-  end
   -- escape special characters for LaTeX
-  captionText = stringEscape(captionText, "latex")
+  captionText = stringEscape(tblCaption, "latex")
   if #tblLabel > 0 then
     captionText = captionText .. " {#" .. tblLabel .. "}"
   end
````

**Comment by cderv** [comment]
@cscheid Sorry I left this one on hold. I switched it to DRAFT because it seems there were more places where we should convert from Markdown instead of `stringify`-ing only. So my first changes only were not enough. I wanted to spend more time to check the different formats. 

I let you finish this. Thank you

**Comment by cscheid** [comment]
We're getting pretty close to a release candidate, and so I'm wary of making large changes (like replacing stringify -> markdown for all cases). I agree it's the right thing to do, but I wonder if we should move this to 1.3.

**Comment by patrickvdb** [comment]
Will anything be done to the use of stringify for LaTeX tables on line 213 in `src/resources/filters/quarto-pre/table-captions.lua`?

**Comment by cderv** [comment]
@cscheid I cleaned the PR and did the change to use Markdown conversion instead of `stringify` only for our raw tables as discussed.

File: tests/docs/smoke-all/2023/01/16/md-captions.qmd [hunk 1]
```diff
@@ -0,0 +1,46 @@
+---
+title: markdown in captions
+format:
+  html: default
+  latex: default
```
Comment (cscheid) [inline tests/docs/smoke-all/2023/01/16/md-captions.qmd:5]: Did you check to see that our test suite supports checking on multiple files at the same time?

What you did here is totally reasonable, and should ideally work. But I don't think we have documents in smoke-all that render to multiple formats, and then test different things on different formats. I suspect this one is only getting tested in HTML.

**Review by cscheid — COMMENTED** [review]
Aside from potentially splitting that `smoke-all` doc into two, this LGTM!

File: tests/docs/smoke-all/2023/01/16/md-captions.qmd (hunk 1 above)
Comment (cscheid) [reply tests/docs/smoke-all/2023/01/16/md-captions.qmd:5]: Yes!

**Review by cscheid — APPROVED** [review]
lgtm

**Comment by cderv** [comment]
Awesome thanks !

**Pull request merged**